_posts/2021-01-04-Designing-REST-services.md

```diff
@@ -0,0 +1,331 @@
+A service contract informs the consumers of the service about the format of the resource (media type), its processing model (protocol) and the links to related resources. 
+The contract also shields the consumer from the implementation details, decreasing the coupling between the service and its consumers, and at the same time heightening security through obfuscation. 
+This loose coupling is further enhanced by applying Postel’s Law: “Be conservative in what you do, be liberal in what you accept from others”, also known as the robustness principle. 
+An example of how to go about implementing this law, is the Tolerant Reader pattern.
+
+{:refdef: style="text-align: center;"}
+<img src="{{ '/img/2021-01-04-Designing-REST-services/servicecontract.png' | prepend: site.baseurl }}" alt="Workspace" class="image" style="margin:0px auto; max-width:100%">
+{: refdef}
+
+Although the service contract has these conventions as a foundation to build upon, there is still a need for some thought to be put into the data modeling of the resources that will be exposed. 
+The decisions on how to divvy up the data assets that make up the business context into exposable resources can be a daunting endeavor, and numerous design factors should be taken into account based on the context in which this services and their contract are designed. 
+
+However, there are some recurring factors influencing the size of the resource representation:
+* Atomicity: Composite resources that share blocks of data can cause the other resource to enter an inconsistent state.
+* Importance of the Information: The optionality of certain components could indicate them belonging to a separate resource representation.
+* Performance and Scalability: the size of the resource and the frequency with which it is accessed, determines how long it takes for it to be passed over the network.
+* Cacheability: This is greatly enhanced if none of the components of the resource changed at a different frequency from the others.
+
+Designing the contract before implementing the service is what is called the contract-first approach. 
+This approach might seem like a lot of design/thinking upfront, but despite this additional preparatory work, it does grant several benefits:
+* Implementation teams that will make use of the service can work in parallel with the service development team.
+* The contract is known before starting development so all teams have an idea of what to expect from the service. This can give a healthy discussion between provider and consumer teams on how to tweak the service an get ahead of mismatches in expectations.
+* An existing contract allows for code generation of stubs and proxies of the service to test for connectivity and availability of the service that will be developed.
+* Contract-first helps with keeping the contract and the underlying implementation loosely coupled as the contract is not based on the code that is run.
+
```

Comment:
Maybe also worth mentioning what the possibles cons are?

Reply:
done